### PR TITLE
Person dialog

### PR DESCRIPTION
**Pull request by Natt913**
Still can't figure out 3/10 of the branches for this class :c

### CODE REVIEW

File: src/AddressBookGUI.java [hunk 1]
```diff
@@ -98,16 +104,18 @@ public AddressBookGUI(AddressBookController controller, AddressBook addressBook)
...
+        saveAsItem.setEnabled(false);
         saveAsItem.addActionListener(e -> {
             final JFileChooser jfc = new JFileChooser();
             if (JFileChooser.APPROVE_OPTION != jfc.showSaveDialog(this)) {
                 return;
             }
             currentFile = jfc.getSelectedFile();
-            if (currentFile == null) {
-                return;
-            }
```
Comment (pvpetit) [inline src/AddressBookGUI.java:110]: Why was this removed?

**Comment by pvpetit** [comment]
Is there no Person dialog test class?

File: src/AddressBookGUI.java (hunk 1 above)
Comment (Natt913) [reply src/AddressBookGUI.java:110]: Okay I didn't make these changes so clearly I fucked something up if the PersonDialogTest class isn't here. It says the file is untracked?

Comment (ZGreening) [reply src/AddressBookGUI.java:110]: PersonDialogTest is untracked. OK, you need to "git add <filepath>" to start tracking the changes of a file. Then "git commit -m "<commit message>". Then "git push --set-upstream origin <yourbranchname>". With the set upstream flag, you will only have to type git push for all future commits to this branch.

Comment (ZGreening) [reply src/AddressBookGUI.java:110]: apparently '<>' dissapear. that should be
git add 'filepath'
git commit -m 'commit message' (inside quotes)
git push --set-upstream origin 'yourbranchname'

File: build.gradle [hunk 2]
```diff
@@ -22,6 +22,7 @@ compileJava {
     }
 }
 
+//Run tests and code coverage tools
 test {
     useJUnitPlatform()
```
Comment (ZGreening) [inline build.gradle:25]: I guess you removed my changes for "compile" to "implementation" for the dependencies? That's fine, but "compile" is deprecated now I believe, so I was just updating how it should look.

File: src/AddressBook.java [hunk 3]
```diff
@@ -52,8 +52,13 @@ public void clear() {
...
             return;
         }
-        fireTableRowsDeleted(0, persons.size() - 1);
+
+        //Clear persons first
+        int lastRow=persons.size()-1;
         persons.clear();
+
+        //Delete table rows
+        fireTableRowsDeleted(0, lastRow);
```
Comment (ZGreening) [inline src/AddressBook.java:61]: This is what the professor was talking about. I don't know if this is your code or @pvpetit, but when we add, edit, or delete, the professor said it is important to check the correct entry was the one adjusted. Not just one less entry exists. This is what I need to change in my own code.

File: src/AddressBookGUI.java [hunk 4]
```diff
@@ -9,25 +9,20 @@
...
-    private final AddressBook addressBook;
-    private final JTable nameList;
-    private final TableRowSorter<AddressBook> tableRowSorter;
+    private final AddressBook addressBook = new AddressBook();;
+    private final AddressBookController controller = new AddressBookController(addressBook);
+    private final JTable nameList = new JTable(addressBook);
+    private final TableRowSorter<AddressBook> tableRowSorter = new TableRowSorter<>(addressBook);;
     private final JButton addButton = new JButton("Add...");
     private final JButton editButton = new JButton("Edit...");
     private final JButton deleteButton = new JButton("Delete");
```
Comment (ZGreening) [inline src/AddressBookGUI.java:28]: I recognize my code! Just leave it, it will get overwritten when I merge mine anyway.

File: src/AddressBookGUI.java [hunk 5]
```diff
@@ -41,14 +36,21 @@ public static void main(String[] args) throws ClassNotFoundException {
...
+        addButton.setName("add");
+        editButton.setName("edit");
+        deleteButton.setName("delete");
+        newItem.setName("new");
+        openItem.setName("open");
+        saveItem.setName("save");
+        saveAsItem.setName("saveAs");
+        printItem.setName("print");
+        quitItem.setName("quit");
+        searchTextField.setName("search");
```
Comment (ZGreening) [inline src/AddressBookGUI.java:51]: My code

File: src/AddressBookGUI.java [hunk 6]
```diff
@@ -68,6 +71,11 @@ public AddressBookGUI(AddressBookController controller, AddressBook addressBook)
         });
         file.add(newItem);
         openItem.addActionListener(e -> {
+            if (saveItem.isEnabled() && JOptionPane.YES_OPTION != JOptionPane.showConfirmDialog(this,
+                    "Are you sure you want to open a different address book? Any unsaved progress will be lost.",
+                    "Open Address Book", JOptionPane.YES_NO_OPTION, JOptionPane.QUESTION_MESSAGE)) {
+                return;
```
Comment (ZGreening) [inline src/AddressBookGUI.java:77]: Missing functionality I added

File: src/AddressBookGUI.java [hunk 7]
```diff
@@ -88,8 +96,6 @@ public AddressBookGUI(AddressBookController controller, AddressBook addressBook)
                 saveAsItem.doClick();
                 return;
             }
-            //FIXME is this needed?
-            //FileSystem fs = new FileSystem();
```
Comment (ZGreening) [inline src/AddressBookGUI.java:92]: Pretty sure I determined this was no longer needed. So this is fine

File: src/AddressBookGUI.java [hunk 8]
```diff
@@ -98,16 +104,18 @@ public AddressBookGUI(AddressBookController controller, AddressBook addressBook)
                         JOptionPane.ERROR_MESSAGE);
             }
         });
+        saveItem.addChangeListener(e -> {
+            // Set saveAsItem with the same state as saveItem
+            saveAsItem.setEnabled(saveItem.isEnabled());
+        });
```
Comment (ZGreening) [inline src/AddressBookGUI.java:110]: SaveAs, matches the state of saveItem now

File: src/FileSystem.java [hunk 9]
```diff
@@ -4,43 +4,81 @@
...
+     * A function to read the data out of an existing AddressBook saved on disk.
+     * Note: this function deletes the data in the current AddressBook.
+     * 
+     * @param addressBook The AddressBook to read the data into.
+     * @param file        The file to read the data from.
+     * @throws SQLException          Thrown if the data in the file is not in the
+     *                               expected SQL format
+     * @throws FileNotFoundException Thrown if the file does not exist or is
+     *                               unreadable
+     */
```
Comment (ZGreening) [inline src/FileSystem.java:17]: All code is supposed to have java docs like this. I haven't got to your files yet, but we need to go through and add comments everywhere!

File: src/PersonDialog.java [hunk 10]
```diff
@@ -30,42 +30,49 @@ public PersonDialog(Frame parent) {
         l = new JLabel("First name:", JLabel.TRAILING);
         p.get().add(l);
         firstName = new JTextField(20);
+        firstName.setName("firstName");
```
Comment (ZGreening) [inline src/PersonDialog.java:31]: I added these for testing.

**Review by ZGreening — CHANGES_REQUESTED** [review]
Yeah I see! This is basically just my own and @pvpetit's code. Figure out how to add it soon. I commented how to from command line somewhere here.

**Comment by ZGreening** [comment]
As you go through and correct things, you can "mark as resolved" on GitHub. That's usually how I keep track of what needs changing. You can also comment back, GitHub sends emails to everyone involved.

File: src/AddressBookGUI.java (hunk 1 above)
Comment (Natt913) [reply src/AddressBookGUI.java:110]: Okay I followed those commands, now what? Another pull request?

Comment (ZGreening) [reply src/AddressBookGUI.java:110]: I don't see your code on here still. Normally it should have updated automatically. You definitely pushed the changes to GitHub? Can you see the changes?

Comment (Natt913) [reply src/AddressBookGUI.java:110]: Everything up-to-date
Branch 'PersonDialog' set up to track remote branch 'PersonDialog' from 'origin'.

'Thats what it says after the push

Comment (ZGreening) [reply src/AddressBookGUI.java:110]: Can you see your code in PersonDialog on GitHub though?

Comment (ZGreening) [reply src/AddressBookGUI.java:110]: Can you see your code in PersonDialog on GitHub though? Try "git status" to double check all important files are tracked.

Comment (Natt913) [reply src/AddressBookGUI.java:110]: No I don't see it on GitHub. Status says its up to date on PersonDialog branch, and 10 commits ahead on master branch

File: tests/PersonDialogTest.java [hunk 11]
```diff
@@ -0,0 +1,204 @@
...
+
+    @Rule
+    private static FrameFixture window = null;
+    public static TemporaryFolder folder = new TemporaryFolder();
+
+
+    @BeforeAll
+    public static void init() {
+        // Prevent program exiting
+        NoExitSecurityManagerInstaller.installNoExitSecurityManager();
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:29]: Not all that important, but I figure I'd point out that I used this because there were lambda functions for exiting the program which needed checking, and That couldn't be done without this line

File: tests/PersonDialogTest.java [hunk 12]
```diff
@@ -0,0 +1,204 @@
...
+
+        // Required for full AssertJ GUI testing
+        FailOnThreadViolationRepaintManager.install();
+    }
+
+    @BeforeEach
+    void setUp() throws IOException, ClassNotFoundException{
+        // Initialize window
+        AddressBookGUI frame = GuiActionRunner.execute(() -> new AddressBookGUI());
```
Comment (ZGreening) [inline tests/PersonDialogTest.java]: You could have initialize directly to PersonDialog, I believe.

Comment (ZGreening) [reply tests/PersonDialogTest.java]: instead of AddressBookGUI. It probably would be better for testing

File: tests/PersonDialogTest.java [hunk 13]
```diff
@@ -0,0 +1,204 @@
...
+    void tearDown() {
+        // Close assertJ window gui
+        window.cleanUp();
+    }
+
+    @AfterAll
+    public static void clean() {
+        // Re-enable program to close after testing completes
+        NoExitSecurityManagerInstaller.installNoExitSecurityManager().uninstall();
+    }
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:53]: This re-allows the program to shutdown properly.

File: tests/PersonDialogTest.java [hunk 14]
```diff
@@ -0,0 +1,204 @@
...
+        dialog.textBox("zip").pressAndReleaseKeys(VK_6, VK_6, VK_6, VK_6, VK_6);
+        dialog.textBox("phone").pressAndReleaseKeys(VK_1, VK_2, VK_3, VK_4, VK_5, VK_6, VK_7, VK_8, VK_9, VK_0);
+
+        // Click 'OK'
+        dialog.button(JButtonMatcher.withText("OK")).click();
+
+        // Test person is added
+        window.table().requireRowCount(1);
+
+    }
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:81]: This looks like it is just a copy of my test. You should be skipping the part about the AddressBookGUI and just testing the PersonDialog. This seems like a great candidate for the mock/stub. Try loading the PersonDialog directly in the @beforeeach function

File: tests/PersonDialogTest.java [hunk 15]
```diff
@@ -0,0 +1,204 @@
...
+            .pressAndReleaseKeys(VK_2, VK_2, VK_4, VK_5, VK_6);
+        
+        dialog.textBox("lastName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_M).releaseKey(VK_SHIFT)
+            .pressAndReleaseKeys(VK_9, VK_7);
+        
+        // Click 'OK'
+        dialog.button(JButtonMatcher.withText("OK")).click();
+
+        // Test person is not added
+        window.table().requireRowCount(0);
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:101]: Add stub/mock. Remove the actual AddressBookGUI, as that is not what you should be testing.

File: tests/PersonDialogTest.java [hunk 16]
```diff
@@ -0,0 +1,204 @@
...
+
+    }
+
+
+    @Test
+    void addPersonNullFirst() {
+        // Test when user exits with OK
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:107]: null and " " are not the same thing. This test could be renamed to addPersonEmptyFirstName()

File: tests/PersonDialogTest.java [hunk 17]
```diff
@@ -0,0 +1,204 @@
...
+        window.table().requireRowCount(0);
+
+    }
+
+
+    @Test
+    void addPersonMixedInvlaid() {
+        // Test first name valid and lastname invalid when user exits with OK
+        window.button("add").click();
+        DialogFixture dialog = window.dialog();
```
Comment (ZGreening) [inline tests/PersonDialogTest.java]: So if you just remove the AddressBookGUI part, you can skip all these clicks to get to the PersonDialog

File: tests/PersonDialogTest.java (hunk 16 above)
Comment (ZGreening) [reply tests/PersonDialogTest.java:107]: This test and the other seem fine, but you need to remove the dependency on the AddressBookGUI for verifying whether the "person" exists or not. If you "mock" the class, you should be able to check the mock to see if the person exists. If that makes sense.

File: tests/PersonDialogTest.java [hunk 18]
```diff
@@ -0,0 +1,204 @@
...
+
+    //     dialog.textBox("firstName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_1).releaseKey(VK_SHIFT)
+    //         .pressAndReleaseKeys(VK_2, VK_2, VK_4, VK_5, VK_6);
+    //     // dialog.textBox("lastName").pressKey(VK_SHIFT).pressAndReleaseKeys(VK_M).releaseKey(VK_SHIFT)
+    //     //     .pressAndReleaseKeys(VK_A, VK_N);
+
+    //     // Click 'OK'
+    //     dialog.button(JButtonMatcher.withText("OK")).click();
+
+    // }
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:141]: Commented out?

File: tests/PersonDialogTest.java [hunk 19]
```diff
@@ -0,0 +1,204 @@
...
+    void addPersonNull() {
+        // Test when user exits with OK
+        window.button("add").click();
+        DialogFixture dialog = window.dialog();
+
+        // Click 'OK'
+        dialog.button(JButtonMatcher.withText("OK")).click();
+
+        // Test person is not added
+        window.table().requireRowCount(0);
```
Comment (ZGreening) [inline tests/PersonDialogTest.java:152]: Same as before, removed addressbookGUI. I am not sure this is null, rather empty strings. I am not sure null would be possible via the GUI. Just rename these to "empty".

File: tests/PersonDialogTest.java [hunk 20]
```diff
@@ -0,0 +1,204 @@
...
+    // }
+
+    // @Test
+    // void nullPerson() {
+    //     AddressBookGUI frame = GuiActionRunner.execute(() -> new AddressBookGUI());
+    //     window = new FrameFixture(frame);
+    //     window.show();
+    //     PersonDialog nullDialog = new PersonDialog(frame, null);
+
+    // }
```
Comment (ZGreening) [inline tests/PersonDialogTest.java]: Commented out code

**Review by ZGreening — CHANGES_REQUESTED** [review]
Biggest thing, we need to remove the dependency of the AddressBookGUI for testing. All of your tests depend on it. Build a mock for it and test in the mock whether or not the items exist in the actual data structure after, not whether they show up on a different GUI.

File: tests/PersonDialogTest.java (hunk 17 above)
Comment (Natt913) [reply tests/PersonDialogTest.java]: Whenever I try to remove the AddressBookGUI part, I get an EDT violation. I am still trying a different way around it though.

File: tests/PersonDialogTest.java (hunk 18 above)
Comment (Natt913) [reply tests/PersonDialogTest.java:141]: This was me trying to increase the branch coverage but it either didn't do anything or didn't work. I left it there to try to come back to it and attempt again. I'll just remove it though.
